third_party/move/move-compiler-v2/src/pipeline/lint_processor/needless_mutable_reference.rs

```diff
@@ -0,0 +1,236 @@
+        use Bytecode::*;
+        use Operation::*;
+        match instr {
+            Assign(_, dst, src, _) => {
+                if self.node_exists(*src) {
+                    self.add_derived_edge(*dst, *src);
+                }
+            },
+            Call(_, dsts, BorrowField(..) | BorrowVariantField(..), srcs, ..) => {
+                debug_assert!(srcs.len() == 1 && dsts.len() == 1);
+                if self.node_exists(srcs[0])
+                    && target.get_local_type(dsts[0]).is_mutable_reference()
+                {
+                    self.add_derived_edge(dsts[0], srcs[0]);
+                }
+            },
+            _ => {},
+        }
+    }
+
+    /// Update mutable usages given `instr`.
+    fn update_mutably_used(&mut self, env: &GlobalEnv, instr: &Bytecode) {
+        use Bytecode::*;
+        use Operation::*;
+        match instr {
+            Call(_, _, WriteRef, srcs, _) => {
+                self.set_and_propagate_mutably_used(srcs[0]);
+            },
+            Call(_, _, Function(mid, fid, _), srcs, _) => {
+                let callee_env = env.get_function_qid(mid.qualified(*fid));
+                callee_env
+                    .get_parameter_types()
+                    .iter()
+                    .enumerate()
+                    .filter(|(_, ty)| ty.is_mutable_reference())
+                    .map(|(i, _)| srcs[i])
+                    .for_each(|src| self.set_and_propagate_mutably_used(src));
+            },
+            _ => {},
+        }
+    }
+
+    /// Add an origin to the tracker.
+    fn add_origin(&mut self, origin: TempIndex, loc: Loc) {
+        self.origins.entry(origin).or_insert(loc);
+        self.derived_edges.entry(origin).or_default();
+    }
+
+    /// Check if a node exists in the tracker.
+    fn node_exists(&self, node: TempIndex) -> bool {
+        self.derived_edges.contains_key(&node)
+    }
+
+    /// Add a derived edge to the tracker.
+    fn add_derived_edge(&mut self, from: TempIndex, to: TempIndex) {
+        self.derived_edges.entry(from).or_default().insert(to);
+        self.propagate_mutably_used(from, to);
+    }
+
+    /// Propagate mutably used information from `from` to `to`.
+    fn propagate_mutably_used(&mut self, from: TempIndex, to: TempIndex) {
+        if self.mutably_used.contains(&from) {
+            self.set_and_propagate_mutably_used(to);
+        }
+    }
+
+    /// Set a mutable reference as mutably used.
+    /// Propagate this information transitively through derived edges.
+    /// Propagation is stopped early if a node is already marked as mutably used.
+    fn set_and_propagate_mutably_used(&mut self, node: TempIndex) {
+        let mut mutably_used = std::mem::take(&mut self.mutably_used);
+        self._set_and_propagate_mutably_used(node, &mut mutably_used);
+        self.mutably_used = mutably_used;
+    }
+
+    /// Helper function for `set_and_propagate_mutably_used`.
+    fn _set_and_propagate_mutably_used(
```

Comment:
Not related to the pr itself but what is the naming convention of this kind of internal helper function? Any examples in the code base?

Reply:
There seem to be some functions in the codebase if you search for `fn _`, but this was not really about adhering to some convention per se.

I have this situation where I need a new internal function that follows a "take, mutate, and set back" pattern due to Rust's ownership rules, so used a similar enough name to the original.

Reply:
In the usual case, you should have a meaningful name on what it specifically does.

But this one is pretty much doing the same thing except for `&self` and `&mut BTreeSet`.  I would probably have called it `..._helper`, so you may find that pattern in code I've written.  To me the `_` prefix means that you may not have any uses of it, so I don't think I'd use that pattern.

You should add a functionl-level comment, that `&self` is lacking the `mutably_used` field, which is passed in separately.  Just to make it easier to understand.